### DISCUSSION TITLE
Is there a rayon parallel implementation for creating array?

### DISCUSSION THREAD

**Discussion by ajlee-kp** [rust-ndarray/ndarray#1435] [Q&A] [answered]
Hello! 

i'm currently creating image array dim of (1, 3, x, y) doing something like this. 
```
        ndarray::Array::from_shape_fn(
            (1, 3, size.width as usize, size.height as usize),
            // BGR -> RGB
            |(_, c, x, y)| (bytes[3 * x + 3 * y * (size.width as usize) + (2 - c)] as f32) / 255., // u8::MAX
        )
```
i'm refactoring iteration and map using rayon feature right now and would love to do that for this ndarray creation step as well.
I'm struggling finding par implementation for ndarray creation step. If you guys can give some ideas that would be awesome.

**Comment by nilgoyette** [comment]
No, not to my knowledge. What you could do is
- init to 0 (should be super fast), then mutate it with the ndarray parallel features.
- if you don't mind using unsafe, I think you can use the uninit feature (then parallel), but I never used it so I prefer not writing possibly false information.

**Reply by bluss** [reply]
Array::uninit is not too bad, the interface is pretty ok, so I think that's a good idea too. But it's also nice to lean back and call `.map_collect()` and let ndarray handle the code that uses uninitialized arrays. :slightly_smiling_face:

**Comment by bluss** [comment, marked as answer]
Zip is for iteration and parallel iteration, and it allows creating arrays.

Maybe you could use `Zip` and `.par_map_collect()` to create the array.
It will only make sense if the element creating computation is heavy enough. Otherwise it will just be overhead and no gain from using rayon.

Zip needs some input to map, you can use indices or a no-allocation array for that?

```rust
Zip::from(ndarray::indices((1, 3, size.width as usize, size.height as usize)))
.par_map_collect(|(i, j, k)| ...)
```
